### ISSUE TITLE
Add tabindex to non-focusable in-page link targets of TOC

### ISSUE THREAD

**Issue by IanPouncey** [gettalong/kramdown#215]
In order for the TOC in-page links to correctly move keyboard focus to the right place in IE and Webkit / Blink browsers, the target heading needs a `tabindex` value. See https://code.google.com/p/chromium/issues/detail?id=262171 for one of multiple duplicate Webkit bug reports.

Without a `tabindex` value the viewport will shift to display the target element, but keyboard focus will remain on the link or move to the top of the document.

To fix this add `tabindex="-1"` to the target heading if it doesn't already have a `tabindex` attribute.

**Comment by gettalong** [comment]
I don't know why this issue is opened here if it is a browser bug.

**Comment by micmath** [comment]
Because disabled people use browsers, some of which require specific markup be used in order that the webpage is accessible. Adding a tabindex to an heading element that you are linking to accomplishes that.

**Comment by IanPouncey** [comment]
This is not a browser bug, although I agree it is odd behaviour.

The solution that the IE and Webkit teams have decided for this problem is that keyboard focus will be moved to an element that has a `tabindex` value (or is an element that is focusable by default).

This has been the case in IE since at least version 6, and Webkit have decided to take the same approach.

**Comment by gettalong** [comment]
I see your point. However, this should probably be done in a post-processor that specifically modifies the document to be accessible. I gather this would just be one of a few changes that would need to be implemented in order to provide complete accessibility.

**Comment by micmath** [comment]
@gettalong I'm not sure I totally follow your point in that last comment because I don't yet know enough about post-processing, but I assume you are not saying that accessibility should be an optional extra part of your software?

(if I may say) @IanPouncey is a very talented expert in web accessibility and it would be a valuable benefit to kramdown to have his contributions to help make your software accessible from its core. I know that many large corporations that might be considering using (and contributing to) any OSS will want to look carefully for the basic accessibility compliance. Would you consider a pull request for that sort of thing or no?

**Comment by gettalong** [comment]
With post-processing I mean modifying the resulting HTML fragment to suit the needs of the application that uses kramdown. Another way would be a custom converter.

Regarding the tabindex: How would the tabindex be specified for each heading? Should we begin at 1 and count up? But what if another part of the page already has such a tabindex? And what if there are other elements between two headings that need a tabindex?

I don't think that all these questions can be answered in a general fashion, or can they? I'm no expert in accessibility.

However, because of all this questions I think that this is probably best done at the application level.

I consider all pull requests, whether I accept them is another thing.

**Comment by micmath** [comment]
I don't see why modifying the resulting HTML fragment wouldn't work. That doesn't appear to how similar functionality (the adding of a linkable id attribute the header elements) is implemented. I would probably base any contribution of an existing example, but only because I'm not very familiar with the code.

The issue that adding a tabindex is meant to solve, if I understand correctly, is that you are intending for people to link to the header elements which the TOC entries refer to, yet no header element is intrinsically focusable -- only a limited set of elements will natively receive focus. The result is that although the browser can scroll the header into view, it does not set keyboard focus on that header (whether that is a bug or not is debatable). So when a keyboard user finishes reading the item and tries to tab away they suddenly fly all the way back up to the top of the page -- very disorienting. By simply adding _any_ positive tabindex to the elements your TOC link to you solve the problem. And so adding tabindex="1" actually is a general solution.

* NB I'm not an accessibility expert either, so hopefully @IanPouncey will confirm my guesses?

**Comment by gettalong** [comment]
But the tabindex does actually have another meaning, too, namely the order in which the tab key moves between elements. So if all headings have a tabindex of 1, this wouldn't work anymore.

I am aware of what adding the tabindex attribute would solve in regards to accessibility but I'm not quite sure if doing this in kramdown itself is the right way to go forward.

**Comment by micmath** [comment]
Being able to tab from one header to the next is possibly a different feature. The point of this issue is to allow tabbing to continue in an _expected_ and natural order, not let the focus to fly all the way back up to the top of the page.

**Comment by IanPouncey** [comment]
If you use a tab index value of `-1` (minus one) it makes the element programatically focusable, with `document.getElementById('foo').focus()` for example, and also allows keyboard focus to be set to it with in page links.

It does not add the element as a tab stop when pressing the tab key, so it is entirely benign as long as it isn't added to an element that is natively focusable.

By making this change to the table of contents as standard it will mean that keyboard users, a group that benefits from in page links like this more than any other, will be able to make use of them in the intended way.

**Comment by gettalong** [comment]
@IanPouncey thanks for the explanation!

One thing is not quite clear, though: Should the tabindex be added to the links in the table of contents or to each heading?

Adding them to the table of contents would be no problem, adding them to the headings (which I guess is the correct thing to do) would require a bit more effort.

**Comment by micmath** [comment]
If it helps to illustrate the request, this is how we are solving the problem at the moment:

```
1. Table of Contents
{:toc #gel-page-toc}

## The Demo
{:tabindex="-1"}

We use a single type hierarchy that is scaled across a range of devices. We make changes to the type sizes depending on the screen size being used and the device capabilities available i.e. touch.
```

And the feature would be that we could get the same effect by just using the following content:

```
1. Table of Contents
{:toc #gel-page-toc}

## The Demo

We use a single type hierarchy that is scaled across a range of devices. We make changes to the type sizes depending on the screen size being used and the device capabilities available i.e. touch.
```

**Comment by gettalong** [comment]
Thanks! I thought that the tabindex would need to be on the heading.

Could there be any negative side effects if the tabindex is added to all headings?

And another question: Are you aware of any other accessibility concerns in the HTML output of kramdown?

**Comment by IanPouncey** [comment]
No, there are no side effects to this. You should check to make sure there isn't already a `tabindex` attribute on the heading as it looks like you do with the `id` attribute (my Ruby-fu is not strong, but that's what it looks like is happening) to make sure that you don't override a defined value. Any value will allow this to work, but `-1` means that you avoid any negative side effects like adding the element to the keyboard tab order.

We've just started using Kramdown, so I haven't seen any other concerns. I'd be happy to raise and help resolve anything I find though.

**Comment by gettalong** [comment]
Thanks for the offer!

I tried to find any site that uses the tabindex attribute in this way but I failed. Even sites like http://www.w3.org/WAI/intro/wcag don't seem to be using the attribute. So I still don't know whether I should add it.

**Comment by IanPouncey** [comment]
It would be nice if the WCAG site was an exemplar for accessibility, but unfortunately it isn't. If you want an example of a website using this technique then the new http://www.paciellogroup.com/ does it for their skip link.

**Comment by gettalong** [comment]
I'm sorry but the website you mention doesn't have any tabindex attribute in its source. I tried the landing page and the "Services" page.

**Comment by micmath** [comment]
Odd, it works for me. Did you actually try tabbing to the skip links, and then following that link by hitting the return key? When I did that I saw that a tabindex was dynamically added to the target element (using javascript I assume). Using Chrome dev tools element inspector I see:

```html
<div class="container" id="content" tabindex="-1">
```

**Comment by IanPouncey** [comment]
It looks like they add it with JavaScript when you tab to the skiplink. Most of the bbc.co.uk sites do the same thing.

It's not necessary to have the JavaScript component though. In this case you are already modifying the markup when adding an `id` to the heading, so the `tabindex` can be added at the same time.

**Comment by gettalong** [comment]
I have decided not to include this for now. Please write a custom converter based on the HTML converter and override `#converter_header` for your use-case.
